cpu/rpx0xx/periph/spi.c

```diff
@@ -0,0 +1,413 @@
+#include "cpu.h"
+#include "mutex.h"
+#include "periph/gpio.h"
+#include "periph/spi.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/**
+ * @brief   Allocate one lock per SPI device.
+ */
+static mutex_t locks[SPI_NUMOF];
+
+/**
+ * @brief   Save the clock prescaler values for faster bus acquisition.
+ */
+typedef struct {
+    spi_clk_t clk;
+    uint8_t cpsdvsr;
+    uint8_t scr;
+} _pl022_clk_t;
+static _pl022_clk_t pl022_clk[SPI_NUMOF];
+
+gpio_t  spi_pin_clk(spi_t spi)
+{
+    assert((unsigned)spi < SPI_NUMOF);
+    return spi_config[spi].clk_pin;
+}
+
+gpio_t  spi_pin_miso(spi_t spi)
+{
+    assert((unsigned)spi < SPI_NUMOF);
+    return spi_config[spi].miso_pin;
+}
+
+gpio_t  spi_pin_mosi(spi_t spi)
+{
+    assert((unsigned)spi < SPI_NUMOF);
+    return spi_config[spi].mosi_pin;
+}
+
+static void _poweron(spi_t spi)
+{
+    uint32_t reset_bit_mask =
+        (spi) ? RESETS_RESET_spi1_Msk : RESETS_RESET_spi0_Msk;
+
+    periph_reset(reset_bit_mask);
+    periph_reset_done(reset_bit_mask);
+}
+
+static void _poweroff(spi_t spi)
+{
+    uint32_t reset_bit_mask =
+        (spi) ? RESETS_RESET_spi1_Msk : RESETS_RESET_spi0_Msk;
+
+    periph_reset(reset_bit_mask);
+}
+
+void spi_init(spi_t spi)
+{
+    DEBUG("[rpx0xx] Call spi_init(spi=%" PRIdFAST8 ")", spi);
+
+    assert((unsigned)spi < SPI_NUMOF);
+
+    /* initialize device lock */
+    mutex_init(&locks[spi]);
+    /* trigger pin initialization */
+    spi_init_pins(spi);
+    /* clock prescaler values must be calculated */
+    pl022_clk[spi].clk = 0xff;
+}
+
+void spi_init_pins(spi_t spi)
+{
+    DEBUG("[rpx0xx] Call spi_init_pins(spi=%" PRIdFAST8 ")", spi);
+
+    assert((unsigned)spi < SPI_NUMOF);
+
+    const gpio_pad_ctrl_t mosi_pad_config = {
+        .drive_strength = DRIVE_STRENGTH_2MA,
+    };
+    const gpio_io_ctrl_t mosi_io_config = {
+        .function_select = FUNCTION_SELECT_SPI,
+    };
+    const gpio_pad_ctrl_t clk_pad_config = {
+        .drive_strength = DRIVE_STRENGTH_2MA,
+    };
+    const gpio_io_ctrl_t clk_io_config = {
+        .function_select = FUNCTION_SELECT_SPI,
+    };
+    const gpio_pad_ctrl_t miso_pad_config = {
+        .pull_up_enable = 1,
+        .input_enable = 1,
+        .schmitt_trig_enable = 1,
+    };
+    const gpio_io_ctrl_t miso_io_config = {
+        .function_select = FUNCTION_SELECT_SPI,
+    };
+
+    if (gpio_is_valid(spi_config[spi].mosi_pin)) {
+        gpio_set_pad_config(spi_config[spi].mosi_pin, mosi_pad_config);
+        gpio_set_io_config(spi_config[spi].mosi_pin, mosi_io_config);
+    }
+    if (gpio_is_valid(spi_config[spi].clk_pin)) {
+        gpio_set_pad_config(spi_config[spi].clk_pin, clk_pad_config);
+        gpio_set_io_config(spi_config[spi].clk_pin, clk_io_config);
+    }
+    if (gpio_is_valid(spi_config[spi].miso_pin)) {
+        gpio_set_pad_config(spi_config[spi].miso_pin, miso_pad_config);
+        gpio_set_io_config(spi_config[spi].miso_pin, miso_io_config);
+    }
+}
+
+void spi_deinit_pins(spi_t spi)
```

Comment:
You also want to lock the SPI mutex here (and unlock in in `spi_init_pins()`) so that any access to the SPI bus will block if the pins are not configured for SPI.

Reply:
Hm, is this really intended behavior? I can see that it would confuse users with unexpected blocking of `spi_acquire()` when they forget to initialize the pins. I think there should be some kind of error message instead.

Also `spi_deinit_pins()` could potentially block, and the docs do not mention this.

Reply:
> Hm, is this really intended behavior?

It is, calling the mutex there is done by all SPI drivers which define this function.
Maybe the documentation isn't good enough regarding this. But it should be improve in another PR.